v2/workloadapi/client_test.go

```diff
@@ -1 +1,210 @@
+	resp := s.makeX509SVIDResponse([]string{"spiffe://example.org/zod"})
+	s.wl.SetX509SVIDResponse(resp)
+
+	w.WaitForUpdates(1)
+	assert.Len(s.T(), w.Errors, 0)
+	assert.Len(s.T(), w.X509SVIDs, 2)
+}
+
+func (s *ClientTestSuite) TestFetchJWTSVID() {
+	spiffeID, _ := spiffeid.New("spiffe://example.org", "mytoken")
+	params := jwtsvid.Params{
+		Subject:        spiffeID,
+		Audience:       "spiffe://example.org/target",
+		ExtraAudiences: []string{"spiffe://example.org/extraAudience"},
+	}
+	svid, err := s.client.FetchJWTSVID(context.TODO(), params)
+
+	require.Nil(s.T(), err)
+	require.NotEmpty(s.T(), svid)
+}
+
+func (s *ClientTestSuite) TestFetchJWTBundles() {
+	spiffeID, _ := spiffeid.New("spiffe://example.org")
+	bundles, err := s.client.FetchJWTBundles(context.TODO())
+	require.Nil(s.T(), err)
+	require.True(s.T(), bundles.Has(spiffeID.TrustDomain()))
+}
+
+func (s *ClientTestSuite) TestWatchJWTBundles() {
+	w := newTestWatcher(s.T())
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	go func() {
+		_ = s.client.WatchJWTBundles(ctx, w)
+	}()
+
+	w.WaitForUpdates(1)
+
+	assert.Len(s.T(), w.Errors, 0)
+	assert.Len(s.T(), w.JwtBundles, 1)
+}
+
+func (s *ClientTestSuite) makeX509SVIDResponse(spiffeIDs []string) *test.X509SVIDResponse {
+	svids := []test.X509SVID{}
+	for _, id := range spiffeIDs {
+		svid, key := s.ca.CreateX509SVID(id)
+		svids = append(svids, test.X509SVID{CertChain: svid, Key: key})
+	}
+	return &test.X509SVIDResponse{
+		Bundle: s.ca.Roots(),
+		SVIDs:  svids,
+	}
+}
+
+func (s *ClientTestSuite) makeJWTSVIDResponse(spiffeID string, audience []string) *workload.JWTSVIDResponse {
+	token := s.ca.CreateJWTSVID(spiffeID, audience)
+	svids := []*workload.JWTSVID{
+		{
+			SpiffeId: spiffeID,
+			Svid:     token,
+		},
+	}
+	return &workload.JWTSVIDResponse{
+		Svids: svids}
+}
+
+type testWatcher struct {
+	t            *testing.T
+	X509SVIDs    []*X509Context
+	JwtBundles   []*jwtbundle.Set
+	Errors       []error
+	updateSignal chan struct{}
+	n            int
+	timeout      time.Duration
+}
+
+func newTestWatcher(t *testing.T) *testWatcher {
+	return &testWatcher{
+		t:            t,
+		updateSignal: make(chan struct{}, 100),
+		timeout:      10 * time.Second,
+	}
+}
+
+func (w *testWatcher) OnX509ContextUpdate(u *X509Context) {
+	w.X509SVIDs = append(w.X509SVIDs, u)
+	w.n++
+	w.updateSignal <- struct{}{}
+}
+
+func (w *testWatcher) OnX509ContextWatchError(err error) {
+	w.Errors = append(w.Errors, err)
+	w.n++
+	w.updateSignal <- struct{}{}
+}
+func (w *testWatcher) OnJWTBundlesUpdate(u *jwtbundle.Set) {
+	w.JwtBundles = append(w.JwtBundles, u)
+	w.n++
+	w.updateSignal <- struct{}{}
+}
+
+func (w *testWatcher) OnJWTBundlesWatchError(err error) {
+	w.Errors = append(w.Errors, err)
+	w.n++
+	w.updateSignal <- struct{}{}
+}
+
+func (w *testWatcher) WaitForUpdates(expectedNumUpdates int) {
+	numUpdates := 0
+	for {
+		select {
+		case <-w.updateSignal:
+			numUpdates++
+		case <-time.After(w.timeout):
```

Comment:
This timer is reset on each iteration... is that intentional?

Reply:
Yes, the idea is to timeout after nothing happened during that time. Do you think it's better to wait a fixed amount, independently of what was going on?

Reply:
It probably won't matter in practice, but in my mind, purpose of the timeout is to keep the test from blocking forever in case something goes wrong. Having one timeout over the whole operation serves that purpose. As written, I feel like people reviewing the code are left why a per-update timeout is important and to me, it wasn't immediately clear. I don't feel too strongly though.